### PR TITLE
Pybind11 coverage

### PR DESCRIPTION
**Pull request by m-chaturvedi**
Run:
`bazel build //bindings/pydrake:generate_pybind_coverage`

in order to produce the `csv` files:
`class_coverage.csv` and `file_coverage.csv`

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/12040)

### CODE REVIEW

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-Lp0Upwk2I81wPu4jgnh:-Lp3tpGr-c4IhEt4p37s:b-80tlqk) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

At a high level, this looks good. I have more low-level comments, but can you take a pass at some clean up first (e.g. get rid of hard-coded constants, like enumerating `attic`, etc.) by Thursday?
</blockquote></details>

Sure.

---

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-Lp0V4F298qwqg-o-We0:-Lp3ttchBIr9jWj7WkLx:bl6vqj2) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

If you have a chance, can you try examining some of the output in a Jupyter notebook, showing the pandas dataframe?
(I have no idea how tractable this is, but would be nice to see.)

See the README here:
https://github.com/RobotLocomotion/drake/blob/master/tools/jupyter/README.md
</blockquote></details>

Yeah, I tried it.  The baobab plot you were talking about are called `sunburst` plots.  I believe they can represent a shallow but wide tree well. 

I tried them using plotly.
https://plot.ly/python/v3/sunburst-charts/

plotly have a js version as well, if we want to go that route.

---

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-Lp0V4F298qwqg-o-We0:-Lp4MI3I-kXcVqIgGWcC:bcmkudm) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Yeah, if you can post a screenshot of what the sunburst looks like (or if it's easy to publish the static HTML with the plot?), that'd be awesome!
And perhaps the relevant notebook?
</blockquote></details>

I had tried it in an adhoc way with a subset of the dataframe, I'm working on it today to do it properly. I'll get it done and send it to you by tomorrow.

---

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-Lp0V4F298qwqg-o-We0:-Lp5Sf8V3XyM0yARqhQF:b-dbvjgm) (no related file):*
<details><summary><i>Previously, m-chaturvedi (Mmanu Chaturvedi) wrote…</i></summary><blockquote>

I had tried it in an adhoc way with a subset of the dataframe, I'm working on it today to do it properly. I'll get it done and send it to you by tomorrow.
</blockquote></details>

If coverage is represented by `A/B`.  This first one is the distribution for `A` and the second one for `B`.






---

**Comment by EricCousineau-TRI** [comment]
Per f2f, distribution is shown well by the sunburst, and hover-over on the arcs can show the percentage.

Would be nice to see if we can overlay that percentage, either via color (blend from green to red for each arc), or perhaps showing a progress bar within in.

**Comment by m-chaturvedi** [comment]
0 % - Green, 100% - Red.

**Comment by m-chaturvedi** [comment]
Here is a folder with a bunch of options: https://ibb.co/album/d26HFa

**Comment by EricCousineau-TRI** [comment]
Sweet! I don't necessarily want to choose a scheme yet, just wanted to know that it was possible.

Can you try out some of the de-hacking to put the attic file list retrievable by a Bazel query? If it takes you more than a couple of hours, can you ping on here?

**Comment by m-chaturvedi** [comment]
Sorry, forgot to let you know.  Updated it yesterday.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot linux-bionic-clang-bazel-experimental-debug please
@drake-jenkins-bot linux-bionic-gcc-bazel-experimental-debug please

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot linux-bionic-unprovisioned-clang-bazel-experimental-debug please
@drake-jenkins-bot linux-bionic-unprovisioned-gcc-bazel-experimental-debug please

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot linux-bionic-unprovisioned-clang-bazel-experimental-debug please
@drake-jenkins-bot linux-bionic-unprovisioned-gcc-bazel-experimental-debug please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by m-chaturvedi** [comment]
@EricCousineau-TRI, sorry about the failures, it builds fine on my development machine.  I will fix them and add the commit.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by EricCousineau-TRI** [comment]
I'm talking to Jeremy briefly to get his opinion on the `attic/` docstring separation.

**Comment by m-chaturvedi** [comment]
___
*[third_party/com_github_pybind_pybind11/mkdoc.py, line 942 at r9](https://reviewable.io/reviews/robotlocomotion/drake/12040#-LuJwJfj91sjqh3vyIA6:-LucfUYwEOb32CiKeAyY:b2nhktp) ([raw file](https://github.com/robotlocomotion/drake/blob/af31adcaf693546f785848a8628cf3307a3d3bca/third_party/com_github_pybind_pybind11/mkdoc.py#L942)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I'm not a huge fan of mutating globals.
I had cleaned this up originally to avoid it, because it was really hard to get the flow of things. (It's still hard to follow, but IMO it's a little less hard than it was before.)

Will see how easy it is to incorporate this change in my PR.
</blockquote></details>

OK.

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-LuDVIu22cKnRp_Q0GoO:-Lucfhod9p3Wmgt0W_2h:bch1687) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Here's my PR:
https://github.com/m-chaturvedi/drake/pull/3

Can you look at it and merge things in?
I merged stuff into `tools/workspace/pybind11`, and melded some stuff. I left some TODOs about updating the test to possibly leverage what's already there.
</blockquote></details>

Just letting you know that I'd merged the changes yesterday.

---

**Comment by m-chaturvedi** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 69 at r9](https://reviewable.io/reviews/robotlocomotion/drake/12040#-LuIaWrz-4tP1xs3tX0X:-LucfpaS7lHKVKDOuRNf:b2nhktp) ([raw file](https://github.com/robotlocomotion/drake/blob/af31adcaf693546f785848a8628cf3307a3d3bca/bindings/pydrake/BUILD.bazel#L69)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

This feels a bit very Drake-specific, and seems asymmetric that we prodouce docstrings for `attic/`, but not coverage artifacts. (Granted, `mkdoc` has a few Drake-specific symbols in it.)

I'm trying to see if there's a slightly better formulation of this in my supplemental PR.
</blockquote></details>

OK.

**Comment by m-chaturvedi** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12040#-LuK-gxgEYGXa4dxQrZe:-LucftERDXwa1rKmg7jF:b2nhktp) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

For organization, I'd prefer that we start out the coverage portions inside `//tools/workspace/pybind`, since that's where the tests for documentation generation live.

I'd also like the coverage generation to just be part of `pybind.bzl` for now.
</blockquote></details>

OK.

---

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please.

**Comment by m-chaturvedi** [comment]
___
*[third_party/com_github_pybind_pybind11/mkdoc.py, line 943 at r13](https://reviewable.io/reviews/robotlocomotion/drake/12040#-LvI2E_G2gu5FDJO5VPc:-LvrLromEJXQEtEBbMaf:b-r319ka) ([raw file](https://github.com/robotlocomotion/drake/blob/8c82e2d4c560afe7ecd097a7fcaa5fdfca7fa473/third_party/com_github_pybind_pybind11/mkdoc.py#L943)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I don't think changing this to recurse would require too  much of a different function length, there should be cheaper ways to shorten this, and I don't think we need a refactoring overhaul to do it, at least for now.
Also, the globals are still there?

If you'd like me to try it (once CI is passing again), let me know.
</blockquote></details>

Oh, I apologize, I think I stashed it and didn't pop it back.

**Comment by m-chaturvedi** [comment]
@jamiesnape:
https://github.com/RobotLocomotion/drake/pull/12718
https://github.com/RobotLocomotion/drake/pull/12727

This can also be PRed separately, if you'd like: https://github.com/RobotLocomotion/drake/compare/master...m-chaturvedi:refactor_mkdoc?expand=1

**Comment by m-chaturvedi** [comment]
___
*[bindings/pydrake/BUILD.bazel, line 271 at r19](https://reviewable.io/reviews/robotlocomotion/drake/12040#-M-o_AVLELhCZL7bLnSA:-M-xFeC2E40o0COmMBWM:b-lk29f1) ([raw file](https://github.com/robotlocomotion/drake/blob/82e1b2c0aaaf544e0dfb880929eaf40e6ca75032/bindings/pydrake/BUILD.bazel#L271)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

BTW Why are we excluding `//bindings/pydrake/visualization`?
</blockquote></details>

@ericcousineau-tri made that change I guesss.

**Comment by m-chaturvedi** [comment]
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-release please
